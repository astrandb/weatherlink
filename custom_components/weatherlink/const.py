"""Constants for the Weatherlink integration."""

from enum import StrEnum

DOMAIN = "weatherlink"
VERSION = "2024.1.1"

MANUFACTURER = "Davis Instruments"
CONFIG_URL = "https://www.weatherlink.com/"

CONF_API_VERSION = "api_version"
CONF_API_KEY_V2 = "api_key_v2"
CONF_API_SECRET = "api_secret"
CONF_API_TOKEN = "apitoken"
CONF_STATION_ID = "station_id"

DISCONNECTED_AFTER_SECONDS = 1830
UNAVAILABLE_AFTER_SECONDS = 3630


class ApiVersion(StrEnum):
    """Supported API versions."""

    API_V1 = "api_v1"
    API_V2 = "api_v2"


class DataKey(StrEnum):
    """Keys for normalized observation data."""

    AQI_VAL = "aqi_val"
    AQI_NOWCAST_VAL = "aqi_nowcast_val"
    BAR_SEA_LEVEL = "bar_sea_level"
    BAR_TREND = "bar_trend"
    DATA_STRUCTURE = "data_structure"
    DEWPOINT = "dewpoint"
    ET_DAY = "et_day"
    ET_MONTH = "et_month"
    ET_YEAR = "et_year"
    HEAT_INDEX = "heat_index"
    HUM = "hum"
    HUM_EXTRA = "hum_extra"
    HUM_IN = "hum_in"
    HUM_OUT = "hum_out"
    MOIST_SOIL = "moist_soil"
    PM_1 = "pm_1"
    PM_2P5 = "pm_2p5"
    PM_10 = "pm_10"
    RAIN_DAY = "rain_day"
    RAIN_MONTH = "rain_month"
    RAIN_RATE = "rain_rate"
    RAIN_STORM = "rain_storm"
    RAIN_STORM_LAST = "rain_storm_last"
    RAIN_STORM_LAST_END = "rain_storm_last_end"
    RAIN_STORM_LAST_START = "rain_storm_last_start"
    RAIN_STORM_START = "rain_storm_start"
    RAIN_YEAR = "rain_year"
    SENSOR_TYPE = "sensor_type"
    SOLAR_PANEL_VOLT = "solar_panel_volt"
    SOLAR_RADIATION = "solar_radiation"
    SUPERCAP_VOLT = "supercap_volt"
    TEMP = "temp"
    TEMP_EXTRA = "temp_extra"
    TEMP_SOIL = "temp_soil"
    TEMP_IN = "temp_in"
    TEMP_OUT = "temp_out"
    TIMESTAMP = "timestamp"
    THW_INDEX = "thw_index"
    THSW_INDEX = "thsw_index"
    TRANS_BATTERY_FLAG = "trans_battery_flag"
    TRANS_BATTERY_VOLT = "trans_battery_volt"
    UUID = "station_id_uuid"
    UV_INDEX = "uv_index"
    WET_BULB = "wet_bulb"
    WET_LEAF = "wet_leaf"
    WIND_CHILL = "wind_chill"
    WIND_DIR = "wind_dir"
    WIND_MPH = "wind_mph"
    WIND_GUST_MPH = "wind_gust_mph"
