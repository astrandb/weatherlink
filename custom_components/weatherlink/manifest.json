{
  "domain": "weatherlink",
  "name": "Weatherlink",
  "codeowners": [
    "@astrandb"
  ],
  "config_flow": true,
  "dependencies": [],
  "documentation": "https://github.com/astrandb/weatherlink",
  "homekit": {},
  "integration_type": "service",
  "iot_class": "cloud_polling",
  "issue_tracker": "https://github.com/astrandb/weatherlink/issues",
  "requirements": [],
  "ssdp": [],
  "version": "0.0.15",
  "zeroconf": []
}
