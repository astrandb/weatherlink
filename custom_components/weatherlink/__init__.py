"""The Weatherlink integration."""
from __future__ import annotations

from datetime import timedelta
import logging

from aiohttp import ClientResponseError
import async_timeout

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .const import (
    CONF_API_KEY_V2,
    CONF_API_SECRET,
    CONF_API_TOKEN,
    CONF_API_VERSION,
    CONF_STATION_ID,
    DOMAIN,
    ApiVersion,
    DataKey,
)
from .pyweatherlink import WLHub, WLHubV2

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]
SENSOR_TYPE_VUE_AND_VANTAGE_PRO = (
    23,
    24,
    27,
    28,
    37,
    43,
    44,
    45,
    46,
    48,
    49,
    50,
    51,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    87,
)

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Weatherlink from a config entry."""

    hass.data.setdefault(DOMAIN, {})
    hass.data[DOMAIN][entry.entry_id] = {}
    if entry.data[CONF_API_VERSION] == ApiVersion.API_V1:
        hass.data[DOMAIN][entry.entry_id]["api"] = WLHub(
            websession=async_get_clientsession(hass),
            username=entry.data[CONF_USERNAME],
            password=entry.data[CONF_PASSWORD],
            apitoken=entry.data[CONF_API_TOKEN],
        )

    if entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
        hass.data[DOMAIN][entry.entry_id]["api"] = WLHubV2(
            websession=async_get_clientsession(hass),
            station_id=entry.data[CONF_STATION_ID],
            api_key_v2=entry.data[CONF_API_KEY_V2],
            api_secret=entry.data[CONF_API_SECRET],
        )
        hass.data[DOMAIN][entry.entry_id]["station_data"] = await hass.data[DOMAIN][
            entry.entry_id
        ]["api"].get_station()

    coordinator = await get_coordinator(hass, entry)
    if not coordinator.last_update_success:
        await coordinator.async_config_entry_first_refresh()
    _LOGGER.debug("First data: %s", coordinator.data)

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok


async def get_coordinator(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> DataUpdateCoordinator:
    """Get the data update coordinator."""
    if "coordinator" in hass.data[DOMAIN][entry.entry_id]:
        return hass.data[DOMAIN][entry.entry_id]["coordinator"]

    def _preprocess(indata: str):
        outdata = {}
        _LOGGER.debug("Received data: %s", indata)
        if entry.data[CONF_API_VERSION] == ApiVersion.API_V1:
            outdata["DID"] = indata["davis_current_observation"].get("DID")
            outdata["station_name"] = indata["davis_current_observation"].get(
                "station_name"
            )
            outdata[DataKey.TEMP_OUT] = indata.get("temp_f")
            outdata[DataKey.TEMP_IN] = indata["davis_current_observation"].get(
                "temp_in_f"
            )
            outdata[DataKey.HUM_IN] = indata["davis_current_observation"].get(
                "relative_humidity_in"
            )
            outdata[DataKey.HUM_OUT] = indata.get("relative_humidity")
            outdata[DataKey.BAR_SEA_LEVEL] = indata.get("pressure_in")
            outdata[DataKey.WIND_MPH] = indata.get("wind_mph")
            outdata[DataKey.WIND_GUST_MPH] = indata["davis_current_observation"].get(
                "wind_ten_min_gust_mph"
            )
            outdata[DataKey.WIND_DIR] = indata.get("wind_degrees")
            outdata[DataKey.DEWPOINT] = indata.get("dewpoint_f")
            outdata[DataKey.RAIN_DAY] = indata["davis_current_observation"].get(
                "rain_day_in"
            )
            outdata[DataKey.RAIN_STORM] = indata["davis_current_observation"].get(
                "rain_storm_in"
            )
            outdata[DataKey.RAIN_RATE] = indata["davis_current_observation"].get(
                "rain_rate_in_per_hr"
            )
            outdata[DataKey.RAIN_MONTH] = indata["davis_current_observation"].get(
                "rain_month_in"
            )
            outdata[DataKey.RAIN_YEAR] = indata["davis_current_observation"].get(
                "rain_year_in"
            )
            outdata[DataKey.BAR_TREND] = indata["davis_current_observation"].get(
                "pressure_tendency_string"
            )

        if entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
            outdata[DataKey.UUID] = indata["station_id_uuid"]
            for sensor in indata["sensors"]:
                # Vue
                if (
                    sensor["sensor_type"] in SENSOR_TYPE_VUE_AND_VANTAGE_PRO
                    and sensor["data_structure_type"] == 10
                ):
                    outdata[DataKey.DATA_STRUCTURE] = sensor["data_structure_type"]
                    outdata[DataKey.TEMP_OUT] = sensor["data"][0]["temp"]
                    outdata[DataKey.HUM_OUT] = sensor["data"][0]["hum"]
                    outdata[DataKey.WIND_MPH] = sensor["data"][0]["wind_speed_last"]
                    outdata[DataKey.WIND_GUST_MPH] = sensor["data"][0][
                        "wind_speed_hi_last_10_min"
                    ]
                    outdata[DataKey.WIND_DIR] = sensor["data"][0]["wind_dir_last"]
                    outdata[DataKey.DEWPOINT] = sensor["data"][0]["dew_point"]
                    outdata[DataKey.RAIN_DAY] = float(
                        sensor["data"][0]["rainfall_daily_in"]
                    )
                    outdata[DataKey.RAIN_STORM] = float(
                        sensor["data"][0]["rain_storm_in"]
                    )
                    outdata[DataKey.RAIN_STORM_START] = sensor["data"][0].get(
                        "rain_storm_start_at"
                    )

                    outdata[DataKey.RAIN_RATE] = sensor["data"][0]["rain_rate_last_in"]
                    outdata[DataKey.RAIN_MONTH] = sensor["data"][0][
                        "rainfall_monthly_in"
                    ]
                    outdata[DataKey.RAIN_YEAR] = sensor["data"][0]["rainfall_year_in"]
                    outdata[DataKey.TRANS_BATTERY_FLAG] = sensor["data"][0][
                        "trans_battery_flag"
                    ]
                if (
                    sensor["sensor_type"] in SENSOR_TYPE_VUE_AND_VANTAGE_PRO
                    and sensor["data_structure_type"] == 2
                ):
                    outdata[DataKey.DATA_STRUCTURE] = sensor["data_structure_type"]
                    outdata[DataKey.TEMP_OUT] = sensor["data"][0]["temp_out"]
                    outdata[DataKey.TEMP_IN] = sensor["data"][0]["temp_in"]
                    outdata[DataKey.BAR_SEA_LEVEL] = sensor["data"][0]["bar"]
                    outdata[DataKey.BAR_TREND] = (
                        float(sensor["data"][0]["bar_trend"]) / 1000
                    )
                    outdata[DataKey.HUM_OUT] = sensor["data"][0]["hum_out"]
                    outdata[DataKey.HUM_IN] = sensor["data"][0]["hum_in"]
                    outdata[DataKey.WIND_MPH] = sensor["data"][0]["wind_speed"]
                    outdata[DataKey.WIND_GUST_MPH] = sensor["data"][0][
                        "wind_gust_10_min"
                    ]
                    outdata[DataKey.WIND_DIR] = sensor["data"][0]["wind_dir"]
                    outdata[DataKey.DEWPOINT] = sensor["data"][0]["dew_point"]
                    outdata[DataKey.RAIN_DAY] = float(sensor["data"][0]["rain_day_in"])
                    outdata[DataKey.RAIN_STORM] = float(
                        sensor["data"][0]["rain_storm_in"]
                    )
                    outdata[DataKey.RAIN_STORM_START] = sensor["data"][0].get(
                        "rain_storm_start_date"
                    )
                    outdata[DataKey.RAIN_RATE] = sensor["data"][0]["rain_rate_in"]
                    outdata[DataKey.RAIN_MONTH] = sensor["data"][0]["rain_month_in"]
                    outdata[DataKey.RAIN_YEAR] = sensor["data"][0]["rain_year_in"]

                if (
                    sensor["sensor_type"] in SENSOR_TYPE_VUE_AND_VANTAGE_PRO
                    and sensor["data_structure_type"] == 23
                ):
                    outdata[DataKey.DATA_STRUCTURE] = sensor["data_structure_type"]
                    outdata[DataKey.TEMP_OUT] = sensor["data"][0]["temp"]
                    outdata[DataKey.HUM_OUT] = sensor["data"][0]["hum"]
                    outdata[DataKey.WIND_MPH] = sensor["data"][0]["wind_speed_last"]
                    outdata[DataKey.WIND_GUST_MPH] = sensor["data"][0][
                        "wind_speed_hi_last_10_min"
                    ]
                    outdata[DataKey.WIND_DIR] = sensor["data"][0]["wind_dir_last"]
                    outdata[DataKey.DEWPOINT] = sensor["data"][0]["dew_point"]
                    outdata[DataKey.RAIN_DAY] = float(
                        sensor["data"][0]["rainfall_day_in"]
                    )
                    x_storm = sensor["data"][0]["rain_storm_current_in"]
                    outdata[DataKey.RAIN_STORM] = float(
                        x_storm if x_storm is not None else 0.0
                    )
                    outdata[DataKey.RAIN_STORM_START] = sensor["data"][0].get(
                        "rain_storm_start_at"
                    )
                    outdata[DataKey.RAIN_RATE] = sensor["data"][0]["rain_rate_last_in"]
                    outdata[DataKey.RAIN_MONTH] = sensor["data"][0]["rainfall_month_in"]
                    outdata[DataKey.RAIN_YEAR] = sensor["data"][0]["rainfall_year_in"]
                    outdata[DataKey.TRANS_BATTERY_FLAG] = sensor["data"][0][
                        "trans_battery_flag"
                    ]
                    outdata[DataKey.TRANS_BATTERY_VOLT] = sensor["data"][0][
                        "trans_battery_volt"
                    ]
                    outdata[DataKey.SUPERCAP_VOLT] = sensor["data"][0]["supercap_volt"]
                    outdata[DataKey.SOLAR_PANEL_VOLT] = sensor["data"][0][
                        "solar_panel_volt"
                    ]

                if sensor["sensor_type"] == 365 and sensor["data_structure_type"] == 21:
                    outdata[DataKey.TEMP_IN] = sensor["data"][0]["temp_in"]
                    outdata[DataKey.HUM_IN] = sensor["data"][0]["hum_in"]
                if sensor["sensor_type"] == 243 and sensor["data_structure_type"] == 12:
                    outdata[DataKey.TEMP_IN] = sensor["data"][0]["temp_in"]
                    outdata[DataKey.HUM_IN] = sensor["data"][0]["hum_in"]
                if sensor["sensor_type"] == 242 and sensor["data_structure_type"] == 12:
                    outdata[DataKey.BAR_SEA_LEVEL] = sensor["data"][0]["bar_sea_level"]
                    outdata[DataKey.BAR_TREND] = sensor["data"][0]["bar_trend"]
                if sensor["sensor_type"] == 242 and sensor["data_structure_type"] == 19:
                    outdata[DataKey.BAR_SEA_LEVEL] = sensor["data"][0]["bar_sea_level"]
                    outdata[DataKey.BAR_TREND] = sensor["data"][0]["bar_trend"]

        return outdata

    async def async_fetch():
        api = hass.data[DOMAIN][entry.entry_id]["api"]
        try:
            async with async_timeout.timeout(10):
                res = await api.request("GET")
                json_data = await res.json()
                hass.data[DOMAIN][entry.entry_id]["current"] = json_data
                return _preprocess(json_data)
        except ClientResponseError as exc:
            _LOGGER.warning("API fetch failed. Status: %s, - %s", exc.code, exc.message)
            raise UpdateFailed(exc) from exc

    hass.data[DOMAIN][entry.entry_id]["coordinator"] = DataUpdateCoordinator(
        hass,
        logging.getLogger(__name__),
        name=DOMAIN,
        update_method=async_fetch,
        update_interval=timedelta(minutes=5),
    )
    await hass.data[DOMAIN][entry.entry_id]["coordinator"].async_refresh()
    return hass.data[DOMAIN][entry.entry_id]["coordinator"]


async def async_migrate_entry(hass, config_entry: ConfigEntry):
    """Migrate old entry."""
    _LOGGER.info("Migrating from version %s", config_entry.version)

    if config_entry.version == 1:
        new_data = {**config_entry.data}

        new_data[CONF_API_VERSION] = ApiVersion.API_V1

        config_entry.version = 2
        hass.config_entries.async_update_entry(config_entry, data=new_data)

    _LOGGER.info("Migration to version %s successful", config_entry.version)

    return True
