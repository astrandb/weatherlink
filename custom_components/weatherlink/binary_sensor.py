"""Platform for binary sensor integration."""
from __future__ import annotations

from dataclasses import dataclass
import logging
from typing import Final

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity import DeviceInfo, EntityCategory
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from . import get_coordinator
from .const import (
    CONF_API_VERSION,
    CONFIG_URL,
    DOMAIN,
    MANUFACTURER,
    ApiVersion,
    DataKey,
)
from .pyweatherlink import WLData

_LOGGER = logging.getLogger(__name__)


@dataclass
class WLBinarySensorDescription(BinarySensorEntityDescription):
    """Class describing Weatherlink binarysensor entities."""

    tag: str | None = None
    exclude_api_ver: set = ()
    exclude_data_structure: set = ()
    aux_sensors: set = ()


SENSOR_TYPES: Final[tuple[WLBinarySensorDescription, ...]] = (
    WLBinarySensorDescription(
        key="TransmitterBattery",
        tag=DataKey.TRANS_BATTERY_FLAG,
        device_class=BinarySensorDeviceClass.BATTERY,
        translation_key="trans_battery",
        entity_category=EntityCategory.DIAGNOSTIC,
        exclude_api_ver=(ApiVersion.API_V1,),
        exclude_data_structure=(2,),
        aux_sensors=(55,),
    ),
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the binary sensor platform."""
    coordinator = await get_coordinator(hass, config_entry)

    entities = [
        WLSensor(coordinator, hass, config_entry, description, 1)
        for description in SENSOR_TYPES
        if (config_entry.data[CONF_API_VERSION] not in description.exclude_api_ver)
        and (
            coordinator.data.get(DataKey.DATA_STRUCTURE)
            not in description.exclude_data_structure
        )
    ]

    aux_entities = []
    if config_entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
        for sensor in hass.data[DOMAIN][config_entry.entry_id]["sensors_metadata"]:
            if sensor["tx_id"] is not None and sensor["tx_id"] > 1:
                aux_entities = [
                    WLSensor(
                        coordinator, hass, config_entry, description, sensor["tx_id"]
                    )
                    for description in SENSOR_TYPES
                    if (sensor["sensor_type"] in description.aux_sensors)
                ]

    async_add_entities(entities + aux_entities)


class WLSensor(CoordinatorEntity, BinarySensorEntity):
    """Representation of a Binary Sensor."""

    entity_description: WLBinarySensorDescription
    sensor_data = WLData()

    def __init__(
        self,
        coordinator,
        hass: HomeAssistant,
        entry: ConfigEntry,
        description: WLBinarySensorDescription,
        tx_id: int,
    ):
        """Initialize the sensor."""
        super().__init__(coordinator)
        self.hass = hass
        self.entry: ConfigEntry = entry
        self.entity_description = description
        self.tx_id = tx_id
        self._attr_has_entity_name = True
        tx_id_part = f"-{self.tx_id}" if self.tx_id > 1 else ""
        self._attr_unique_id = (
            f"{self.get_unique_id_base()}{tx_id_part}-{self.entity_description.key}"
        )
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, f"{self.get_unique_id_base()}{tx_id_part}")},
            name=self.generate_name(),
            manufacturer=MANUFACTURER,
            model=self.generate_model(),
            sw_version=self.get_firmware(),
            serial_number=self.get_serial(),
            configuration_url=CONFIG_URL,
            via_device=(DOMAIN, self.get_unique_id_base())
            if tx_id_part != ""
            else None,
        )

    def get_unique_id_base(self):
        """Generate base for unique_id."""
        unique_base = None
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V1:
            unique_base = self.coordinator.data["DID"]
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
            unique_base = self.coordinator.data[DataKey.UUID]
        return unique_base

    def get_firmware(self) -> str | None:
        """Get firmware version."""
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
            return self.hass.data[DOMAIN][self.entry.entry_id]["station_data"][
                "stations"
            ][0].get("firmware_version")
        return None

    def get_serial(self) -> str | None:
        """Get serial number."""
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
            return self.hass.data[DOMAIN][self.entry.entry_id]["station_data"][
                "stations"
            ][0].get("gateway_id_hex")
        return None

    def generate_name(self):
        """Generate device name."""
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V1:
            return self.coordinator.data["station_name"]
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
            if self.tx_id == 1:
                return self.hass.data[DOMAIN][self.entry.entry_id]["station_data"][
                    "stations"
                ][0]["station_name"]
            for sensor in self.hass.data[DOMAIN][self.entry.entry_id][
                "sensors_metadata"
            ]:
                if sensor["sensor_type"] in (55, 56) and sensor["tx_id"] == self.tx_id:
                    return f"{sensor['product_name']} ID{sensor['tx_id']}"

        return "Unknown devicename"

    def generate_model(self):
        """Generate model string."""
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V1:
            return "WeatherLink - API V1"
        if self.entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
            model: str = self.hass.data[DOMAIN][self.entry.entry_id]["station_data"][
                "stations"
            ][0].get("product_number")
        if model == "6555":
            return f"WeatherLinkIP {model}"
        if model.startswith("6100"):
            return f"WeatherLink Live {model}"
        if model.startswith("6313"):
            return f"WeatherLink Console {model}"
        return "WeatherLink"

    @property
    def is_on(self):
        """Return the state of the sensor."""
        # _LOGGER.debug("Key: %s", self.entity_description.key)
        return self.coordinator.data[self.tx_id].get(self.entity_description.tag)
