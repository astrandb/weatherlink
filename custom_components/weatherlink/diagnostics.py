"""Diagnostics support for Weatherlink."""
from __future__ import annotations

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .const import (
    CONF_API_KEY_V2,
    CONF_API_SECRET,
    CONF_API_TOKEN,
    CONF_API_VERSION,
    DOMAIN,
    ApiVersion,
)

TO_REDACT = {
    CONF_PASSWORD,
    CONF_USERNAME,
    CONF_API_TOKEN,
    CONF_API_SECRET,
    CONF_API_KEY_V2,
    "user_email",
}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict:
    """Return diagnostics for a config entry."""
    coordinator: DataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id][
        "coordinator"
    ]
    station_data = hass.data[DOMAIN][config_entry.entry_id].get("station_data", {})
    current = hass.data[DOMAIN][config_entry.entry_id]["current"]
    sensor_metadata = hass.data[DOMAIN][config_entry.entry_id]["sensors_metadata"]

    sensor_data = {}
    if config_entry.data[CONF_API_VERSION] == ApiVersion.API_V2:
        sensor_data = await hass.data[DOMAIN][config_entry.entry_id][
            "api"
        ].get_all_sensors()

    diagnostics_data = {
        "info": async_redact_data(config_entry.data, TO_REDACT),
        "station_data": async_redact_data(station_data, TO_REDACT),
        "all_sensor_data": async_redact_data(sensor_data, TO_REDACT),
        "sensor_metadata": async_redact_data(sensor_metadata, TO_REDACT),
        "current_data": async_redact_data(current, TO_REDACT),
        "data": async_redact_data(coordinator.data, TO_REDACT),
    }

    return diagnostics_data
