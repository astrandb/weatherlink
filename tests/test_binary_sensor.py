"""Provide tests for weatherlink sensors."""

from unittest.mock import patch

import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    snapshot_platform,
)
from syrupy import SnapshotAssertion

from custom_components.weatherlink.const import DOMAIN
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from . import setup_integration
from .const import ENTRY_ID, MOCK_CONFIG_V2


@pytest.mark.usefixtures("entity_registry_enabled_by_default")
async def test_binary_sensor(
    hass: HomeAssistant,
    bypass_get_data,
    bypass_get_station,
    bypass_get_all_sensors,
    snapshot: SnapshotAssertion,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test sensor states."""

    mock_config_entry = MockConfigEntry(
        domain=DOMAIN, version=2, data=MOCK_CONFIG_V2, entry_id=ENTRY_ID
    )

    with patch("custom_components.weatherlink.PLATFORMS", [Platform.BINARY_SENSOR]):
        await setup_integration(hass, mock_config_entry)

    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)


# @pytest.mark.parametrize(
#     "data_file_name", ["station_166.json", "station_135.json", "station_88.json"]
# )
# @pytest.mark.usefixtures("entity_registry_enabled_by_default")
# async def test_sensor_additional_stations(
#     hass: HomeAssistant,
#     bypass_get_data,
#     snapshot: SnapshotAssertion,
#     entity_registry: er.EntityRegistry,
# ) -> None:
#     """Test states for additional stations."""

#     mock_config_entry = MockConfigEntry(
#         domain=DOMAIN, data=MOCK_CONFIG, entry_id=ENTRY_ID
#     )

#     with patch("custom_components.viva.PLATFORMS", [Platform.SENSOR]):
#         await setup_integration(hass, mock_config_entry)

#     await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
