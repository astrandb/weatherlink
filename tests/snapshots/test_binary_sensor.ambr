# serializer version: 1
# name: test_binary_sensor[binary_sensor.strp81_connectivity-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.strp81_connectivity',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.CONNECTIVITY: 'connectivity'>,
    'original_icon': None,
    'original_name': 'Connectivity',
    'platform': 'weatherlink',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'timestamp',
    'unique_id': '03e7585a-4f29-4e7c-b6cb-d9e17313b07c-Timestamp',
    'unit_of_measurement': None,
  })
# ---
# name: test_binary_sensor[binary_sensor.strp81_connectivity-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'connectivity',
      'friendly_name': 'Strp81 Connectivity',
      'last_update': datetime.datetime(2024, 12, 28, 11, 55, tzinfo=datetime.timezone.utc),
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.strp81_connectivity',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_binary_sensor[binary_sensor.strp81_transmitter_battery-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.strp81_transmitter_battery',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.BATTERY: 'battery'>,
    'original_icon': None,
    'original_name': 'Transmitter battery',
    'platform': 'weatherlink',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'trans_battery',
    'unique_id': '03e7585a-4f29-4e7c-b6cb-d9e17313b07c-TransmitterBattery',
    'unit_of_measurement': None,
  })
# ---
# name: test_binary_sensor[binary_sensor.strp81_transmitter_battery-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'battery',
      'friendly_name': 'Strp81 Transmitter battery',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.strp81_transmitter_battery',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
